### PR TITLE
New Keithley 2200 series power supplies

### PR DESCRIPTION
**Pull request by ashokbruno**

### CODE REVIEW

File: tests/instruments/test_all_instruments.py [hunk 1]
```diff
@@ -111,6 +111,7 @@ def test_adapter_arg(cls):
     "HP6634A",
     "HP8657B",
     "Keithley2000",
+    "Keithley2200",
```
Comment (bmoneke) [inline tests/instruments/test_all_instruments.py:114]: Don't add your instrument to the list of exceptions (instruments not yet changed), but make you instrument conform to the tests (see comment at init).
That means, revert this change.

File: tests/instruments/keithley/test_keithley2200.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+def test_set_voltage_current():
+    # instr.ch1.voltage should produce the following commands
+    with expected_protocol(
+        Keithley2200,
+        [(b"INST:SEL CH1;VOLT 1.456", None)],
+    ) as instr:
+        instr.ch1.voltage = 1.456
+
+    with expected_protocol(
```
Comment (bmoneke) [inline tests/instruments/keithley/test_keithley2200.py:46]: Please put every test (expected protocol and assert) in its own test method. That way you always know which method is broken and not just that something is broken

File: tests/instruments/keithley/test_keithley2200.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        Keithley2200,
+        [("INST:SEL CH1;VOLT?", 1.456)],
+    ) as inst:
+        val = inst.ch1.voltage
```
Comment (bmoneke) [inline tests/instruments/keithley/test_keithley2200.py:50]: You can combine it in one line: `assert Inst.value == 5`.

File: pymeasure/instruments/keithley/keithley2200.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+class Channel():
+    """ Implementation of a Keithley 2200 channel. """
+
+    VOLTAGE_RANGE = [0, 70]
+    CURRENT_RANGE = [0, 45]
+
+    output_enabled = Instrument.control(
+        "SOURCE:OUTP:ENAB?", "SOURCE:OUTP:ENAB %d",
+        """A boolean property that controls whether the output is enabled, takes
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:41]: Please adjust the docstrings according to the current [documentation](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings).

**Review by bmoneke — CHANGES_REQUESTED** [review]
Thanks for your contribution. It is great to see new contributors help maintaining the code base by providing new instruments.

I left a few comments, especially regarding recent changes (new channel implementation, obligatory name parameter and docstrings).

File: pymeasure/instruments/keithley/keithley2200.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    output_enabled = Instrument.control(
+        "SOURCE:OUTP:ENAB?", "SOURCE:OUTP:ENAB %d",
+        """ A boolean property representing output state.""",
+        validator=strict_discrete_set,
+        values={True: 1, False: 0},
+        map_values=True
+    )
+
+    voltage = Instrument.control(
```
Comment (bilderbuchi) [inline pymeasure/instruments/keithley/keithley2200.py:47]: We have started to standardise naming of power supply properties, could you adapt your properties?
https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#power-supplies (`voltage` and `current` as measurements (i.e. get-only), and `voltage_setpoint` and `current_limit` as settings or controls (depending if you can read those back from this device).

File: pymeasure/instruments/keithley/keithley2200.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        "MEAS:CURR?",
+        """ A measurement property that reads the measured current in Amps.""",
+    )
+
+    measure_voltage = Instrument.measurement(
+        "MEAS:VOLT?",
+        """ A measurement property that reads the measured current in Volts.""",
+    )
+
+    measure_power = Instrument.measurement(
```
Comment (bilderbuchi) [inline pymeasure/instruments/keithley/keithley2200.py:71]: "measure" is a verb, so such a name fits a method better. For a property, `power` would be preferable. If the docstring says `"""Measure the power in Watt (float).""", that would follow the [guidance](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings) for docstrings, too. (Also for other properties)

File: pymeasure/instruments/keithley/keithley2200.py [hunk 7]
```diff
@@ -23,129 +23,93 @@
...
-        self.instrument.write(f"INST:SEL CH{self.number};{command}")
+    def insert_id(self, command):
+        return f"INST:SEL CH{self.id};{command}"
 
 
 class Keithley2200(Instrument):
     """ Represents the Keithley 2230 Power Supply.
     """
+    def __init__(self, adapter, name="Keithley2200"):
+        super().__init__(adapter, name)
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:99]: Please keep the kwargs in:
```suggestion
    def __init__(self, adapter, name="Keithley2200", **kwargs):
        super().__init__(adapter, name, **kwargs)
```

File: pymeasure/instruments/keithley/keithley2200.py (hunk 6 above)
Comment (bmoneke) [reply pymeasure/instruments/keithley/keithley2200.py:71]: > that would follow the [guidance](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings) for docstrings, too. (Also for other properties)

As I also pointed out.
Please have a look at the docstrings in [adding instruments](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html), how the property docstrings should look like.

**Review by bmoneke — COMMENTED** [review]
Thanks for the changes.

The channel implementation and changed tests look good, only the docstrings (as bilderbuchi pointed out) remain to polish.

File: pymeasure/instruments/keithley/keithley2200.py (hunk 7 above)
Comment (ashokbruno) [reply pymeasure/instruments/keithley/keithley2200.py:99]: fixed in commit

File: pymeasure/instruments/keithley/keithley2200.py (hunk 5 above)
Comment (bmoneke) [reply pymeasure/instruments/keithley/keithley2200.py:47]: The voltage you set (this property) should be called `voltage_setpoint`.

What you call "measure_voltage" should be the actually measured voltage, as @bilderbuchi pointed out.

Please don't resolve a conversation, if you did not implement the requested change.

File: pymeasure/instruments/keithley/keithley2200.py (hunk 6 above)
Comment (bmoneke) [reply pymeasure/instruments/keithley/keithley2200.py:71]: Here again, please adjust the docstrings of your properties to the new style:

Start with "Set", "Measure", or "Control" (depending on the property type). You might add the type and limits in parentheses afterwards:

 For example "Measure current power output in Watts" or (for the property above) "Control the output voltage in Volts (float between 0 and 10)."

File: pymeasure/instruments/keithley/keithley2200.py (hunk 7 above)
Comment (bmoneke) [reply pymeasure/instruments/keithley/keithley2200.py:99]: Did you push that commit, because they are not in the current version online.

File: tests/instruments/keithley/test_keithley2200.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+def test_init():
+    with expected_protocol(
+        Keithley2200,
+        [],
+    ):
+        pass  # Verify the expected communication.
+
+
+def test_set_voltage():
```
Comment (bmoneke) [inline tests/instruments/keithley/test_keithley2200.py:38]:
```suggestion
def test_voltage():
```
You test the setter and getter, don't you?

File: tests/instruments/keithley/test_keithley2200.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        ],
+    ) as instr:
+        instr.ch_1.voltage = 1.456
+        assert instr.ch_1.voltage == 1.456
+        assert instr.ch_1.measure_voltage == 1.456
+        instr.ch_3.voltage = 1.456
+
+
+def test_set_current():
+    # instr.ch1.voltage should produce the following commands
```
Comment (bmoneke) [inline tests/instruments/keithley/test_keithley2200.py:56]: Also here, you should not use "Set".
This comment is out of place. You don't need it anyways.
```suggestion
```

**Review by bmoneke — COMMENTED** [review]
Hey @ashokbruno could it be, that you forgot to push a commit, because you resolved conversations, which are not yet resolved in the public code.

File: pymeasure/instruments/keithley/keithley2200.py (hunk 7 above)
Comment (ashokbruno) [reply pymeasure/instruments/keithley/keithley2200.py]: my mistake, pushed the commit.

File: pymeasure/instruments/keithley/keithley2200.py [hunk 10]
```diff
@@ -31,85 +31,92 @@
...
+        map_values=True,
     )
 
-    voltage = Instrument.control(
-        "VOLT?", "VOLT %g",
-        """ A floating point property representing the output voltage in Volts.""",
+    voltage_setpoint = Instrument.control(
+        "VOLT?",
+        "VOLT %g",
+        """ Set output voltage in Volts.""",
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:51]: You can read the output setpoint as well, therefore "control":
```suggestion
        """ Control output voltage in Volts.""",
```

File: pymeasure/instruments/keithley/keithley2200.py [hunk 11]
```diff
@@ -31,85 +31,92 @@
...
+        values=VOLTAGE_RANGE,
     )
 
-    current = Instrument.control(
-        "CURR?", "CURR %g",
-        """ A floating point property that representing the output current in Amps.""",
+    current_limit = Instrument.control(
+        "CURR?",
+        "CURR %g",
+        """ Set output current in Amps.""",
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:59]:
```suggestion
        """ Control output current in Amps.""",
```

File: pymeasure/instruments/keithley/keithley2200.py [hunk 12]
```diff
@@ -31,85 +31,92 @@
...
 
     VOLTAGE_RANGE = [0, 70]
     CURRENT_RANGE = [0, 45]
 
     output_enabled = Instrument.control(
-        "SOURCE:OUTP:ENAB?", "SOURCE:OUTP:ENAB %d",
-        """ A boolean property representing output state.""",
+        "SOURCE:OUTP:ENAB?",
+        "SOURCE:OUTP:ENAB %d",
+        """ Enables or Disables the output.""",
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:42]:
```suggestion
        """ Control the output state.""",
```

File: pymeasure/instruments/keithley/keithley2200.py [hunk 13]
```diff
@@ -31,85 +31,92 @@
...
+        "MEAS:POW?",
+        """ Measure the power in watts.""",
     )
 
     voltage_limit = Instrument.control(
-        "VOLT:LIM?", "VOLT:LIM %g",
-        """ A property which represents the maximum voltage limit.""",
+        "VOLT:LIM?",
+        "VOLT:LIM %g",
+        """ Limits the maximum voltage that can be set.""",
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:82]:
```suggestion
        """ Control the maximum voltage that can be set.""",
```

File: pymeasure/instruments/keithley/keithley2200.py [hunk 14]
```diff
@@ -31,85 +31,92 @@
...
-        values=VOLTAGE_RANGE
+        values=VOLTAGE_RANGE,
     )
 
     voltage_limit_enabled = Instrument.control(
-        "VOLT:LIM:STAT?", "VOLT:LIM:STAT %d",
-        """A boolean property that representing voltage limit.""",
+        "VOLT:LIM:STAT?",
+        "VOLT:LIM:STAT %d",
+        """ Turns on or off the maximum voltage limit.""",
```
Comment (bmoneke) [inline pymeasure/instruments/keithley/keithley2200.py:90]:
```suggestion
        """ Control whether the maximum voltage limit is enabled.""",
```

**Review by bmoneke — APPROVED** [review]
Thanks for that commit.
I suggested some small changes in the docstring formulation (you might commit them as they are).
With these, we can merge, if you want to.

File: pymeasure/instruments/keithley/keithley2200.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        values={True: 1, False: 0},
+        map_values=True,
+    )
+
+    def insert_id(self, command):
+        return f"INST:SEL CH{self.id};{command}"
+
+
+class Keithley2200(Instrument):
+    """Represents the Keithley 2230 Power Supply."""
```
Comment (evancharlton) [inline pymeasure/instruments/keithley/keithley2200.py:101]: **question**: Is this a copy/paste bug?

I would've thought this should say:
```suggestion
    """Represents the Keithley 2200 Power Supply."""
```

**Comment by bmoneke** [comment]
Do you need help with the conflict resolution?
You have to
1. merge the pymeasure master into your branch.
2. fix the `AUTHORS.txt` file
3. commit and push

**Comment by bmoneke** [comment]
You have to add
`Robert Roos` to the end of `AUTHORS.txt` to resolve the conflict.

**Comment by bmoneke** [comment]
Thanks for your work @ashokbruno
